demos/http/http_demo_s3_download_multithreaded/http_demo_s3_download_multithreaded.c

```diff
@@ -0,0 +1,1022 @@
+
+static QueueOpStatus_t retrieveHTTPResponse( mqd_t responseQueue,
+                                             ResponseItem_t * responseItem )
+{
+    QueueOpStatus_t returnStatus = QUEUE_OP_SUCCESS;
+
+    /* Return value of mq_receive. */
+    int mqread = 0;
+
+    /* Read response from queue. */
+    mqread = mq_receive( responseQueue, ( char * ) responseItem,
+                         sizeof( ResponseItem_t ), NULL );
+    responseItem->response.pBuffer = responseItem->responseBuffer;
+
+    if( mqread == -1 )
+    {
+        if( errno != EAGAIN )
+        {
+            /* Error other than due to not blocking. */
+            LogError( ( "Failed to read from response queue with error %s.",
+                        strerror( errno ) ) );
+            returnStatus = QUEUE_OP_FAILURE;
+        }
+        else
+        {
+            returnStatus = QUEUE_OP_WOULD_BLOCK;
+        }
+    }
+
+    if( returnStatus == QUEUE_OP_SUCCESS )
+    {
+        if( mqread != sizeof( ResponseItem_t ) )
+        {
+            LogError( ( "Response from response queue has incorrect size." ) );
+            returnStatus = QUEUE_OP_FAILURE;
+        }
+    }
+
+    return returnStatus;
+}
+
+/*-----------------------------------------------------------*/
+
+static bool getS3ObjectFileSize( const HTTPRequestInfo_t * requestInfo,
+                                 mqd_t requestQueue,
+                                 mqd_t responseQueue,
+                                 size_t * pFileSize )
+{
+    bool returnStatus = true;
+    HTTPStatus_t httpStatus = HTTP_SUCCESS;
+    QueueOpStatus_t queueOpStatus = QUEUE_OP_SUCCESS;
+
+    /* Request data sent over queue. */
+    ResponseItem_t responseItem = { 0 };
+
+    /* The location of the file size in contentRangeValStr. */
+    char * pFileSizeStr = NULL;
+
+    /* String to store the Content-Range header value. */
+    char * contentRangeValStr = NULL;
+    size_t contentRangeValStrLength = 0;
+
+    LogInfo( ( "Getting file object size from host..." ) );
+
+    /* Request bytes 0 to 0. S3 will respond with a Content-Range
+     * header that contains the size of the file in it. This header will look
+     * like: "Content-Range: bytes 0-0/FILESIZE". The body will have a single
+     * byte that we are ignoring. */
+    queueOpStatus = requestS3ObjectRange( requestInfo,
+                                          requestQueue,
+                                          0,
+                                          0 );
+
+    if( queueOpStatus != QUEUE_OP_SUCCESS )
+    {
+        returnStatus = false;
+    }
+
+    if( returnStatus == true )
+    {
+        do
+        {
+            queueOpStatus = retrieveHTTPResponse( responseQueue, &responseItem );
+        } while ( queueOpStatus == QUEUE_OP_WOULD_BLOCK );
+
+        if( queueOpStatus != QUEUE_OP_SUCCESS )
+        {
+            returnStatus = false;
+        }
+    }
+
+    if( returnStatus == true )
```

Comment:
Please also check that a 206 status code is in the responseItem.response.statusCode before doing other stuff in the function.